agent/config/builder.go

```diff
@@ -371,9 +371,9 @@ func (b *Builder) Build() (rt RuntimeConfig, err error) {
 	// derive other advertise addresses from the advertise address
 	advertiseAddrLAN := b.makeIPAddr(b.expandFirstIP("advertise_addr", c.AdvertiseAddrLAN), advertiseAddr)
 	advertiseAddrWAN := b.makeIPAddr(b.expandFirstIP("advertise_addr_wan", c.AdvertiseAddrWAN), advertiseAddrLAN)
-	rpcAdvertiseAddr := b.makeTCPAddr(b.expandFirstIP("advertise_addresses.rpc", c.AdvertiseAddrs.RPC), advertiseAddrLAN, serverPort)
-	serfAdvertiseAddrLAN := b.makeTCPAddr(b.expandFirstIP("advertise_addresses.serf_lan", c.AdvertiseAddrs.SerfLAN), advertiseAddrLAN, serfPortLAN)
-	serfAdvertiseAddrWAN := b.makeTCPAddr(b.expandFirstIP("advertise_addresses.serf_wan", c.AdvertiseAddrs.SerfWAN), advertiseAddrWAN, serfPortWAN)
+	rpcAdvertiseAddr := &net.TCPAddr{IP: advertiseAddrLAN.IP, Port: serverPort}
+	serfAdvertiseAddrLAN := &net.TCPAddr{IP: advertiseAddrLAN.IP, Port: serfPortLAN}
+	serfAdvertiseAddrWAN := &net.TCPAddr{IP: advertiseAddrWAN.IP, Port: serfPortWAN}
 
```

Comment:
If am right, we still need to reconcile this part - https://github.com/hashicorp/consul/blob/bfa70a10cac3876960cbe0e3d7a7bb482fe2876a/agent/agent.go#L606 

In the above, when setting up memberlist we use `config.SerfPortLAN` and `config.SerfPortWAN` as the port memberlist actually uses to bind to. Since you've already ensured that serfAdvertiseAddrLAN/WAN has the right port, its less confusing to use that directly in that line, instead of assuming that serfPortLAN and serfAdvertiseAddrLAN.Port will end up being the same value because of these lines.

Reply:
I thought I had already fixed this. The `SerfPortLAN/WAN` fields are somewhat leftovers from when we were getting things to work and the addresses weren't `*net.TCPAddr` fields yet. They are used in `command/agent.go` and `agent/agent_endpoint_test.go`

Reply:
I've changed the code in `agent/agent.go` to use the ports from the derived addresses. Getting rid of the other port values means to get rid of the notion that there is only one port per protocol.

Reply:
I

Reply:
would be great if we could remove serfportlan/wan from the runtime config entirely but that might be even more invasive, feel free to discard this suggestion if that's not feasible

Reply:
I've started and stopped this a couple of times. Lets do this in a separate commit. There is nothing stopping us to do this at a later stage.